In [1]:
# import de la table de fait
import pandas as pd
commandes = pd.read_csv("/Users/emmanuelfabre/Documents/Professionel/2025.09_Jedha - Data analyst/09_29 - Fullstack data/3_Projets/6.4_Kiosque à Pizza/2025.12.07-Base fictive/commandes_detaillees_generees_v2.csv", header=0)
commandes.head()

,num_commande,kiosque_id,date_commande,source,client_id,type_ligne,id_article,quantite,prix_unitaire_ttc,prix_unitaire_ht,montant_tva,taux_tva,id_remise,moyen_paiement
0,CMD-248750,k01,2022-07-25 18:36:00,CAISSE,NaN,PRODUIT,PIZ069,1,14.5,13.18,1.32,0.1,NaN,Ticket Resto
1,CMD-248701,k01,2022-07-25 20:03:00,CAISSE,c01031,PRODUIT,PIZ071,1,13.9,12.64,1.26,0.1,NaN,Espèces
2,CMD-248701,k01,2022-07-25 20:03:00,CAISSE,c01031,SUPPLEMENT,SUP038,1,0.5,0.45,0.05,0.1,NaN,Carte Bancaire
3,CMD-248767,k01,2022-08-31 11:16:00,CAISSE,c01312,PRODUIT,PIZ002,1,0.0,0.00,0.00,0.1,REM003,Ticket Resto
4,CMD-248801,k01,2022-08-31 13:15:00,CAISSE,c01112,PRODUIT,PIZ099,1,14.5,13.18,1.32,0.1,NaN,Carte Bancaire


In [2]:
commandes.sample(10)

,num_commande,kiosque_id,date_commande,source,client_id,type_ligne,id_article,quantite,prix_unitaire_ttc,prix_unitaire_ht,montant_tva,taux_tva,id_remise,moyen_paiement
179808,CMD-396161,k04,2025-01-09 12:10:00,CAISSE,NaN,PRODUIT,PIZ020,1,15.9,14.45,1.45,0.1,NaN,Carte Bancaire
246032,CMD-441510,k04,2025-10-05 13:37:00,INTERNET,c00956,SUPPLEMENT,SUP028,1,0.5,0.45,0.05,0.1,NaN,En Ligne
13072,CMD-251302,k01,2022-12-01 13:23:00,CAISSE,NaN,PRODUIT,PIZ040,1,16.5,15.00,1.50,0.1,NaN,Carte Bancaire
25686,CMD-253130,k01,2023-02-12 18:01:00,CAISSE,c00767,SUPPLEMENT,SUP015,1,0.5,0.45,0.05,0.1,NaN,Ticket Resto
222876,CMD-316027,k01,2025-07-13 19:11:00,CAISSE,NaN,SUPPLEMENT,SUP004,1,0.5,0.45,0.05,0.1,NaN,Carte Bancaire
24022,CMD-353199,k04,2023-02-03 21:46:00,CAISSE,c00785,PRODUIT,PIZ004,1,14.9,13.55,1.35,0.1,NaN,Espèces
68450,CMD-259532,k01,2023-08-28 22:12:00,CAISSE,c00738,PRODUIT,PIZ025,1,12.5,11.36,1.14,0.1,NaN,Carte Bancaire
109607,CMD-120044,k02,2024-03-08 21:27:00,CAISSE,c01027,PRODUIT,PIZ040,1,16.5,15.00,1.50,0.1,NaN,Carte Bancaire
2229,CMD-100150,k02,2022-10-05 13:21:00,CAISSE,NaN,PRODUIT,PIZ010,1,14.9,13.55,1.35,0.1,NaN,Carte Bancaire
115153,CMD-122305,k02,2024-03-31 13:35:00,CAISSE,c01005,PRODUIT,PIZ052,1,15.9,14.45,1.45,0.1,NaN,Carte Bancaire


In [3]:
description = commandes.describe(include='all')
description

,num_commande,kiosque_id,date_commande,source,client_id,type_ligne,id_article,quantite,prix_unitaire_ttc,prix_unitaire_ht,montant_tva,taux_tva,id_remise,moyen_paiement
count,261492,261492,261492,261492,198025,261492,261492,261492.0,261492.000000,261492.000000,261492.000000,2.614920e+05,16464,261492
unique,123134,4,109030,4,1177,2,225,NaN,NaN,NaN,NaN,NaN,7,4
top,CMD-207355,k01,2024-09-28 22:34:00,CAISSE,c01324,PRODUIT,PIZ059,NaN,NaN,NaN,NaN,NaN,REM005,Carte Bancaire
freq,7,87771,16,211344,9933,228360,2184,NaN,NaN,NaN,NaN,NaN,2592,126782
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,10.727239,9.750905,0.976334,1.000000e-01,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,5.784312,5.259026,0.525295,3.825420e-13,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,0.000000,0.000000,0.000000,1.000000e-01,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,5.900000,5.360000,0.540000,1.000000e-01,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,13.900000,12.640000,1.260000,1.000000e-01,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,14.500000,13.180000,1.320000,1.000000e-01,NaN,NaN


In [4]:
type_col = commandes.dtypes
type_col

num_commande          object
kiosque_id            object
date_commande         object
source                object
client_id             object
type_ligne            object
id_article            object
quantite               int64
prix_unitaire_ttc    float64
prix_unitaire_ht     float64
montant_tva          float64
taux_tva             float64
id_remise             object
moyen_paiement        object
dtype: object

In [5]:
#Création des colonnes supplémentaires pour l'EDA évolution du CA
commandes['date_commande'] = pd.to_datetime(commandes['date_commande'])
commandes['mois'] = commandes['date_commande'].dt.month
commandes['annee'] = commandes['date_commande'].dt.year
commandes['annee_mois'] = commandes['date_commande'].dt.strftime('%Y-%m')
commandes["jour_semaine"] = commandes["date_commande"].dt.day_name()
commandes['revenu'] = commandes['quantite'] * commandes['prix_unitaire_ttc']


### le gérant constate une baisse du chiffre d'affaire =>
- confirmer et quantifier la baisse du CA (nb commandes, montant et panier moyen)
- identifier les kiosques et les sources de commandes concernés
- identifier le upselling/cross-selling
- voir l'évolution de la consommation des 20% de clients qui génère 80% du CA 

In [6]:
#calcul du revenu total
revenu_total = commandes["revenu"].sum()
revenu_total

np.float64(2805087.2500000005)

In [7]:
#revenu total par kiosque
revenu_kiosque = commandes.groupby("kiosque_id")["revenu"].sum()
revenu_kiosque

kiosque_id
k01    955652.25
k02    870313.15
k03    148242.15
k04    830879.70
Name: revenu, dtype: float64

In [8]:
liste_kap = commandes['kiosque_id'].unique()
liste_kap

array(['k01', 'k04', 'k02', 'k03'], dtype=object)

### vérifier la date de commande la plus ancienne pour chaque kiosque pour s'assurer de la comparabilité

In [9]:
# 1. Conversion des dates en datetime
commandes['date_commande'] = pd.to_datetime(commandes['date_commande'], errors='coerce')

# 2. Fonction de tri spécifique pour les IDs au format K01, K02, etc.
def trier_kiosque(id):
    """Extrait le numéro du kiosque pour le tri (ex: K01 -> 1)"""
    try:
        # Extraire le numéro et le convertir en entier
        return int(id[1:])  # On prend tout après le 'K' et on convertit en int
    except:
        return 999  # Valeur par défaut pour les IDs non conformes

# 3. Tri de la liste des kiosques
liste_kap_sorted = sorted(liste_kap, key=trier_kiosque)

# 4. Affichage des résultats triés
print("Dates de premières et dernières commandes par kiosque :")
print("=" * 80)

for kiosque_id in liste_kap_sorted:
    # Filtrer les commandes pour ce kiosque
    kiosque_commands = commandes[commandes['kiosque_id'] == kiosque_id]

    # Calculer les dates min/max
    first_date = kiosque_commands['date_commande'].min()
    last_date = kiosque_commands['date_commande'].max()

    # Calculer la durée d'activité
    if pd.notna(first_date) and pd.notna(last_date):
        duration = (last_date - first_date).days
        duration_str = f"{duration} jours"
    else:
        duration_str = "N/A"

    # Affichage formaté
    print(f"{kiosque_id}:")
    print(f"  Première commande: {first_date.strftime('%d/%m/%Y') if pd.notna(first_date) else 'N/A'}")
    print(f"  Dernière commande:  {last_date.strftime('%d/%m/%Y') if pd.notna(last_date) else 'N/A'}")
    print(f"  Durée d'activité:   {duration_str}")
    print("-" * 80)


Dates de premières et dernières commandes par kiosque :
k01:
  Première commande: 25/07/2022
  Dernière commande:  07/12/2025
  Durée d'activité:   1231 jours
--------------------------------------------------------------------------------
k02:
  Première commande: 01/10/2022
  Dernière commande:  08/12/2025
  Durée d'activité:   1164 jours
--------------------------------------------------------------------------------
k03:
  Première commande: 16/05/2025
  Dernière commande:  07/12/2025
  Durée d'activité:   205 jours
--------------------------------------------------------------------------------
k04:
  Première commande: 01/10/2022
  Dernière commande:  07/12/2025
  Durée d'activité:   1163 jours
--------------------------------------------------------------------------------


In [10]:
# CA (revenu) par jour de semaine
ordre_jours = [
    "Monday", "Tuesday", "Wednesday", "Thursday", 
    "Friday", "Saturday", "Sunday"]

commandes["jour_semaine"] = pd.Categorical(
    commandes["jour_semaine"],
    categories=ordre_jours,
    ordered=True
)

revenu_jour_semaine = commandes.groupby("jour_semaine")["revenu"].sum()
revenu_jour_semaine

/var/folders/bk/9frx0d7n4tb3lvy4_t2cb_xc0000gn/T/ipykernel_50369/2492501482.py:12: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  revenu_jour_semaine = commandes.groupby("jour_semaine")["revenu"].sum()


jour_semaine
Monday       282027.20
Tuesday      322777.90
Wednesday    330901.00
Thursday     357802.75
Friday       621406.30
Saturday     511437.10
Sunday       378735.00
Name: revenu, dtype: float64

In [11]:
# analyse par saison pour évaluer la bonne performance de 2024 (effet JO ?)
import numpy as np

conditions = [
    # Hiver (Décembre, Janvier, Février)
    (commandes['mois'].isin([12, 1, 2])),
    # Printemps (Mars, Avril, Mai)
    (commandes['mois'].isin([3, 4, 5])),
    # Été (Juin, Juillet, Août)
    (commandes['mois'].isin([6, 7, 8])),
    # Automne (Septembre, Octobre, Novembre)
    (commandes['mois'].isin([9, 10, 11]))
]

# 3. Définition des valeurs correspondantes (les noms des saisons)
valeurs_saisons = ['Hiver', 'Printemps', 'Été', 'Automne']

# 4. Application de np.select() pour créer la colonne 'saison'
# Le paramètre default='Inconnu' est facultatif mais recommandé
commandes['saison'] = np.select(conditions, valeurs_saisons, default='Inconnu')

# Affichage des 10 premières lignes pour vérification
commandes[['date_commande', 'mois', 'saison']].head(10)

,date_commande,mois,saison
0,2022-07-25 18:36:00,7,Été
1,2022-07-25 20:03:00,7,Été
2,2022-07-25 20:03:00,7,Été
3,2022-08-31 11:16:00,8,Été
4,2022-08-31 13:15:00,8,Été
5,2022-08-31 13:15:00,8,Été
6,2022-08-31 19:37:00,8,Été
7,2022-08-31 19:37:00,8,Été
8,2022-08-31 19:37:00,8,Été
9,2022-08-31 22:32:00,8,Été


!!! La base de données va d'octobre 2022 à début décembre 2025.
Pour comparer les années entre elle, nous avons retenu pour 2023, 2024 et 2025 une période d'analyse du 1er janvier au 30 novembre

In [12]:
# création de la colonne "periode_analyse" avec un tag du 1er janvier au 30 novembre de chaque année car il manque décembre 2025
# et exclusion du kiosque k03 pour ne conserver que les kiosques "historiques"
cond_annee = commandes['annee'].isin([2023, 2024, 2025])
cond_mois = commandes['mois'].between(1, 11)
cond_kiosque = commandes['kiosque_id'] != 'k03'
commandes_filtrees = cond_annee & cond_mois & cond_kiosque
commandes['periode_analyse'] = commandes_filtrees.astype(int)

In [13]:
#somme du CA selon le tag periode_analyse
revenu_periode = commandes.groupby("periode_analyse")["revenu"].sum()
revenu_periode

periode_analyse
0     500523.05
1    2304564.20
Name: revenu, dtype: float64

In [14]:
#création du dataframe sur la période concernée
commandes_periode_analyse = commandes[commandes['periode_analyse'] == 1]

# EVOLUTION DU CHIFFRE D'AFFAIRE

In [15]:
# analyse de l'évolution du CA par année
revenu_annee = commandes_periode_analyse.groupby("annee")["revenu"].sum()
print("Chiffre d'affaire annuel :")
display(pd.DataFrame(revenu_annee))

# analyse de l'évolution du nombre de commandes par année
nbcom_annee = commandes_periode_analyse.groupby("annee")["num_commande"].nunique()
print("Nombre annuel des commandes :")
display(pd.DataFrame(nbcom_annee))

#analyse du panier moyen
panier_moyen_annee = revenu_annee / nbcom_annee
print("Panier Moyen Annuel :")
display(pd.DataFrame(panier_moyen_annee))


Chiffre d'affaire annuel :


,revenu
annee,
2023,767056.95
2024,815915.90
2025,721591.35


Nombre annuel des commandes :


,num_commande
annee,
2023,32330
2024,36788
2025,32917


Panier Moyen Annuel :


,0
annee,
2023,23.725857
2024,22.178860
2025,21.921541


In [16]:
# analyse de l'évolution du CA par année et saison
revenu_annee_saison = commandes_periode_analyse.groupby(["saison", "annee"])["revenu"].sum()
#print("Chiffre d'affaire annuel par saison :")
#print(revenu_annee_saison)

# analyse de l'évolution du nombre de commandes par année et saison
nbcom_annee_saison = commandes_periode_analyse.groupby(["saison", "annee"])["num_commande"].nunique()
print("Nombre annuel des commandes par saison :")
display(pd.DataFrame(nbcom_annee_saison))

#analyse du panier moyen et saison
panier_moyen_saison = revenu_annee_saison / nbcom_annee_saison
#print("Panier Moyen Annuel par saison :")
#print(panier_moyen_saison)

Nombre annuel des commandes par saison :


num_commande
saison    annee              
Automne   2023           9243
          2024           9352
          2025           8402
Hiver     2023           4784
          2024           5769
          2025           5350
Printemps 2023           8752
          2024          10982
          2025           9781
Été       2023           9551
          2024          10685
          2025           9384

In [17]:
kiosque_filtre_id = "k04"
commandes_kiosque = commandes_periode_analyse[commandes_periode_analyse['kiosque_id'] == kiosque_filtre_id]

# 1. Calcul du revenu annuel (déjà fait dans votre code précédent)
revenu_annee_kiosque_filtre = commandes_kiosque.groupby("annee")["revenu"].sum().reset_index()
revenu_annee_kiosque_filtre = revenu_annee_kiosque_filtre.rename(columns={
    "annee": "Année",
    "revenu": "Revenu"
})

# 2. Nombre de commandes par année (format numérique)
nbcom_annee_kiosque_filtre = commandes_kiosque.groupby("annee")["num_commande"].nunique().reset_index()
nbcom_annee_kiosque_filtre = nbcom_annee_kiosque_filtre.rename(columns={
    "annee": "Année",
    "num_commande": "Nombre de commandes"
})

# 3. Calcul du panier moyen
panier_annee_kiosque_filtre = (revenu_annee_kiosque_filtre.set_index('Année')['Revenu'] /
                              nbcom_annee_kiosque_filtre.set_index('Année')['Nombre de commandes']).reset_index()
panier_annee_kiosque_filtre = panier_annee_kiosque_filtre.rename(columns={"index": "Année", 0: "Panier moyen"})

# Fonction pour formater les valeurs en euros
def format_euro(x):
    if pd.isna(x):
        return "0,00 €"
    return f"{x:,.2f} €".replace(",", " ").replace(".", ",")

# Fonction pour formater les nombres (sans décimales)
def format_nombre(x):
    if pd.isna(x):
        return "0"
    return f"{x:,.0f}".replace(",", " ")

# 4. Création des DataFrames finaux
# Nombre de commandes (format numérique)
nbcom_df = nbcom_annee_kiosque_filtre.copy()
nbcom_df['Nombre de commandes'] = nbcom_df['Nombre de commandes'].apply(format_nombre)

# Panier moyen (format monétaire)
panier_df = panier_annee_kiosque_filtre.copy()
panier_df['Panier moyen (€)'] = panier_df['Panier moyen'].apply(format_euro)

# Revenu annuel
revenu_df = revenu_annee_kiosque_filtre.copy()
revenu_df['Revenu (€)'] = revenu_df['Revenu'].apply(format_euro)
revenu_df = revenu_df[['Année', 'Revenu (€)']]  # Réorganisation des colonnes

# 5. Calcul des totaux
total_commandes = nbcom_annee_kiosque_filtre['Nombre de commandes'].sum()
total_panier = (revenu_annee_kiosque_filtre['Revenu'].sum() /
               nbcom_annee_kiosque_filtre['Nombre de commandes'].sum())


# 6. Ajout des lignes de total
# Pour le nombre de commandes
total_com_row = pd.DataFrame({
    'Année': ['TOTAL'],
    'Nombre de commandes': [format_nombre(total_commandes)]
})

# Pour le panier moyen
total_panier_row = pd.DataFrame({
    'Année': ['TOTAL'],
    'Panier moyen (€)': [format_euro(total_panier)]
})

# Pour le revenu annuel
total_revenu = revenu_annee_kiosque_filtre['Revenu'].sum()
total_row = pd.DataFrame({
    'Année': ['TOTAL'],
    'Revenu (€)': [format_euro(total_revenu)]
})

# 7. Concatenation avec les totaux
nbcom_df = pd.concat([
    nbcom_df[['Année', 'Nombre de commandes']],
    total_com_row
], ignore_index=True)

panier_df = pd.concat([
    panier_df[['Année', 'Panier moyen (€)']].rename(columns={'Panier moyen': 'Panier moyen (€)'}),
    total_panier_row
], ignore_index=True)


revenu_df = pd.concat([
    revenu_df, total_row], ignore_index=True)

# 8. Affichage des résultats
print(f"\n{'='*60}")
print(f"STATISTIQUES POUR LE KIOSQUE {kiosque_filtre_id.upper()}")
print(f"{'='*60}")

print("\nNombre annuel de commandes:")
display(nbcom_df)

print("\nPanier moyen annuel:")
display(panier_df)

print("\nRevenu annuel:")
display(revenu_df[['Année', 'Revenu (€)']])



STATISTIQUES POUR LE KIOSQUE K04

Nombre annuel de commandes:


,Année,Nombre de commandes
0,2023,10 101
1,2024,11 431
2,2025,10 912
3,TOTAL,32 444



Panier moyen annuel:


,Année,Panier moyen (€)
0,2023,"23,69 €"
1,2024,"21,87 €"
2,2025,"21,52 €"
3,TOTAL,"22,32 €"



Revenu annuel:


,Année,Revenu (€)
0,2023,"239 342,75 €"
1,2024,"249 999,15 €"
2,2025,"234 849,35 €"
3,TOTAL,"724 191,25 €"


### Conclusion pour les 3 kiosques "historiques" : 
- augmentation du CA entre 2023 et 2024 au global puis baisse entre 2024 et 2025 
- diminution du CA entre 2024 et 2025 dans les 3 autres kiosques (k01 = -16%, K02 = -12%, K04 = -6% et global = -12%)
- la baisse du CA observée entre 2024 et 2025 est liée à une baisse du nombre de commandes et du panier moyen (-2 euros en moyenne)
- 


In [18]:
# 1. Fonctions de formatage
def format_euro(x):
    """Formate un nombre en euros avec le format français"""
    if pd.isna(x) or x == 0:
        return "0,00 €"
    return f"{x:,.2f} €".replace(",", " ").replace(".", ",")

def format_nombre(x):
    """Formate un nombre avec séparateurs de milliers"""
    if pd.isna(x):
        return "0"
    return f"{x:,.0f}".replace(",", " ")

# 2. Chiffre d'affaires annuel par source
print("\n" + "="*60)
print("CHIFFRE D'AFFAIRES ANNUEL PAR SOURCE")
print("="*60)

revenu_par_source = commandes_periode_analyse.groupby(["source", "annee"])["revenu"].sum().reset_index()
revenu_par_source["Revenu (€)"] = revenu_par_source["revenu"].apply(format_euro)
revenu_pivot = revenu_par_source.pivot(index="annee", columns="source", values="Revenu (€)")
display(revenu_pivot)

# 3. Nombre de commandes par source
print("\n" + "="*60)
print("NOMBRE DE COMMANDES PAR SOURCE")
print("="*60)

nb_com_par_source = commandes_periode_analyse.groupby(["source", "annee"])["num_commande"].nunique().reset_index()
nb_com_par_source["Nombre de commandes"] = nb_com_par_source["num_commande"].apply(format_nombre)
nb_com_pivot = nb_com_par_source.pivot(index="annee", columns="source", values="Nombre de commandes")
display(nb_com_pivot)

# 4. Panier moyen annuel par source
print("\n" + "="*60)
print("PANIER MOYEN ANNUEL PAR SOURCE")
print("="*60)

# Calcul du panier moyen
panier_moyen = (
    commandes_periode_analyse.groupby(["source", "annee"])["revenu"].sum() /
    commandes_periode_analyse.groupby(["source", "annee"])["num_commande"].nunique()
).reset_index(name="panier_moyen")

panier_moyen["Panier moyen (€)"] = panier_moyen["panier_moyen"].apply(format_euro)
panier_pivot = panier_moyen.pivot(index="annee", columns="source", values="Panier moyen (€)")
display(panier_pivot)

# 5. Version synthétique complète
print("\n" + "="*80)
print("SYNTHÈSE COMPLÈTE PAR SOURCE ET ANNÉE")
print("="*80)

# Création d'un DataFrame complet
synthese = commandes_periode_analyse.groupby(["source", "annee"]).agg({
    "revenu": "sum",
    "num_commande": "nunique"
}).reset_index()

synthese["panier_moyen"] = synthese["revenu"] / synthese["num_commande"]

# Formatage des colonnes
synthese["Revenu (€)"] = synthese["revenu"].apply(format_euro)
synthese["Nombre de commandes"] = synthese["num_commande"].apply(format_nombre)
synthese["Panier moyen (€)"] = synthese["panier_moyen"].apply(format_euro)

# Affichage par source
for source in synthese["source"].unique():
    print(f"\nSource: {source}")
    source_data = synthese[synthese["source"] == source]
    display(source_data[["annee", "Revenu (€)", "Nombre de commandes", "Panier moyen (€)"]])



CHIFFRE D'AFFAIRES ANNUEL PAR SOURCE


source,CAISSE,DELIVEROO,INTERNET,UBER
annee,,,,
2023,"642 298,15 €","14 364,40 €","46 723,20 €","63 671,20 €"
2024,"651 835,75 €","20 951,10 €","93 599,65 €","49 529,40 €"
2025,"563 393,40 €","28 523,20 €","85 084,60 €","44 590,15 €"



NOMBRE DE COMMANDES PAR SOURCE


source,CAISSE,DELIVEROO,INTERNET,UBER
annee,,,,
2023,25 479,920,2 710,3 221
2024,27 976,1 333,4 749,2 730
2025,24 262,1 757,4 426,2 472



PANIER MOYEN ANNUEL PAR SOURCE


source,CAISSE,DELIVEROO,INTERNET,UBER
annee,,,,
2023,"25,21 €","15,61 €","17,24 €","19,77 €"
2024,"23,30 €","15,72 €","19,71 €","18,14 €"
2025,"23,22 €","16,23 €","19,22 €","18,04 €"



SYNTHÈSE COMPLÈTE PAR SOURCE ET ANNÉE

Source: CAISSE


,annee,Revenu (€),Nombre de commandes,Panier moyen (€)
0,2023,"642 298,15 €",25 479,"25,21 €"
1,2024,"651 835,75 €",27 976,"23,30 €"
2,2025,"563 393,40 €",24 262,"23,22 €"



Source: DELIVEROO


,annee,Revenu (€),Nombre de commandes,Panier moyen (€)
3,2023,"14 364,40 €",920,"15,61 €"
4,2024,"20 951,10 €",1 333,"15,72 €"
5,2025,"28 523,20 €",1 757,"16,23 €"



Source: INTERNET


,annee,Revenu (€),Nombre de commandes,Panier moyen (€)
6,2023,"46 723,20 €",2 710,"17,24 €"
7,2024,"93 599,65 €",4 749,"19,71 €"
8,2025,"85 084,60 €",4 426,"19,22 €"



Source: UBER


,annee,Revenu (€),Nombre de commandes,Panier moyen (€)
9,2023,"63 671,20 €",3 221,"19,77 €"
10,2024,"49 529,40 €",2 730,"18,14 €"
11,2025,"44 590,15 €",2 472,"18,04 €"


### Conclusions après croisement avec la source :
- Azay le rideau : baisse des commandes et panier moyen en CAISSE, augmentation du panier moyen sur internet
- La riche : baisse des commandes en CAISSE + UBER, baisse du panier moyen en CAISSE et UBER, hausse en DELIVEROO et INTERNET
- Parcay : baisse des commandes en CAISSE + UBER, baisse du panier moyen en CAISSE et UBER, hausse en DELIVEROO et INTERNET

# UPSELLING ET CROSS-SELLING

In [19]:
commandes_periode_analyse['is_supplement'] = commandes_periode_analyse['id_article'].astype(str).str.startswith('SUP')

import numpy as np

# Colonne 'ca_supplement' : le revenu si c'est un SUP, 0 sinon
commandes_periode_analyse['ca_supplement'] = np.where(
    commandes_periode_analyse['is_supplement'] == True,
    commandes_periode_analyse['revenu'],
    0
)

sum_suppl = commandes_periode_analyse["ca_supplement"].sum()
print(sum_suppl)

13662.5


/var/folders/bk/9frx0d7n4tb3lvy4_t2cb_xc0000gn/T/ipykernel_50369/973866946.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  commandes_periode_analyse['is_supplement'] = commandes_periode_analyse['id_article'].astype(str).str.startswith('SUP')
/var/folders/bk/9frx0d7n4tb3lvy4_t2cb_xc0000gn/T/ipykernel_50369/973866946.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  commandes_periode_analyse['ca_supplement'] = np.where(


In [20]:
# analyse du upselling

part_suppl = ["source", "annee"]
suppl_ca = commandes_periode_analyse.groupby(part_suppl).agg(
    ca_global=("revenu", "sum"),
    ca_supplement=("ca_supplement", "sum"),
)
suppl_ca["part_suppl_ratio"] = (suppl_ca['ca_supplement'] / suppl_ca['ca_global'] * 100).round(2)

print("--- CA Global et CA Suppléments agrégés par Année")
print(suppl_ca)


--- CA Global et CA Suppléments agrégés par Année
                 ca_global  ca_supplement  part_suppl_ratio
source    annee                                            
CAISSE    2023   642298.15         3760.5              0.59
          2024   651835.75         3859.0              0.59
          2025   563393.40         3260.5              0.58
DELIVEROO 2023    14364.40           92.0              0.64
          2024    20951.10          138.5              0.66
          2025    28523.20          182.5              0.64
INTERNET  2023    46723.20          288.5              0.62
          2024    93599.65          558.0              0.60
          2025    85084.60          526.0              0.62
UBER      2023    63671.20          414.5              0.65
          2024    49529.40          314.5              0.63
          2025    44590.15          268.0              0.60


In [21]:
# analyse du cross_selling (Boissons et Desserts)
cond_boi = commandes_periode_analyse['id_article'].astype(str).str.startswith("BOI")
cond_des = commandes_periode_analyse['id_article'].astype(str).str.startswith("DES")
commandes_periode_analyse['is_cross'] = cond_boi | cond_des

# Colonne 'ca_cross' : le revenu si c'est un SUP, 0 sinon
commandes_periode_analyse['ca_cross'] = np.where(
    commandes_periode_analyse['is_cross'] == True,
    commandes_periode_analyse['revenu'],
    0
)

sum_cross = commandes_periode_analyse["ca_cross"].sum()
print(sum_cross)

20295.000000000004


/var/folders/bk/9frx0d7n4tb3lvy4_t2cb_xc0000gn/T/ipykernel_50369/605865804.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  commandes_periode_analyse['is_cross'] = cond_boi | cond_des
/var/folders/bk/9frx0d7n4tb3lvy4_t2cb_xc0000gn/T/ipykernel_50369/605865804.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  commandes_periode_analyse['ca_cross'] = np.where(


In [22]:
part_cross = ["source", "annee"]
cross_ca = commandes_periode_analyse.groupby(part_cross).agg(
    ca_cross_global=("revenu", "sum"),
    ca_cross=("ca_cross", "sum"),
)
cross_ca["part_cross_ratio"] = (cross_ca['ca_cross'] / cross_ca['ca_cross_global'] * 100).round(2)

print("--- CA Global et CA Suppléments agrégés par Année")
print(cross_ca)


--- CA Global et CA Suppléments agrégés par Année
                 ca_cross_global  ca_cross  part_cross_ratio
source    annee                                             
CAISSE    2023         642298.15   4382.70              0.68
          2024         651835.75   5496.30              0.84
          2025         563393.40   4764.55              0.85
DELIVEROO 2023          14364.40    218.45              1.52
          2024          20951.10    388.45              1.85
          2025          28523.20    349.10              1.22
INTERNET  2023          46723.20    719.15              1.54
          2024          93599.65    993.00              1.06
          2025          85084.60    944.85              1.11
UBER      2023          63671.20    727.70              1.14
          2024          49529.40    659.00              1.33
          2025          44590.15    651.75              1.46


### Conclusions : 
- un upselling stable entre 2023 et 2025
- la part du cross-selling dans le CA est moins importante en CAISSE que les autres sources
- un cross-selling stable ou en légère progression
=> voir la ventilation des pizzas selon la sous-catégorie pour comprendre la baisse du panier moyen

# Sous-catégories des pizzas

In [23]:
import openpyxl
base_produits = pd.read_excel("/Users/emmanuelfabre/Documents/Professionel/2025.09_Jedha - Data analyst/09_29 - Fullstack data/3_Projets/6.4_Kiosque à Pizza/2025.12.07-Base fictive/base_produits.xlsx")
print("Aperçu de la base produits :")
print(base_produits.head())

Aperçu de la base produits :
  id_produit       nom_produit taille_pizza categorie_produit  \
0     PIZ001   2 TOMATES 29 cm        29 cm            Pizzas   
1     PIZ002   2 TOMATES 33 cm        33 cm            Pizzas   
2     PIZ003   4 FROMAGES 29CM        29 cm            Pizzas   
3     PIZ004   4 FROMAGES 33CM        33 cm            Pizzas   
4     PIZ005  6 FROMAGES 29 cm        29 cm            Pizzas   

  ss_categorie_produit  prix_produit  
0           Classiques          12.5  
1           Classiques          14.5  
2           Gourmandes          12.9  
3           Gourmandes          14.9  
4           Gourmandes          12.9  


In [24]:
commandes_enrichies = pd.merge(
    commandes_periode_analyse,
    base_produits[["id_produit", "categorie_produit", "ss_categorie_produit"]],
    left_on="id_article",
     right_on = "id_produit",
    how="left"
)

# Vérification : assurez-vous que la colonne 'ss_categorie_produit' est bien remplie
print("\nAperçu des commandes enrichies :")
print(commandes_enrichies[['id_article', 'categorie_produit', 'ss_categorie_produit', 'revenu']].head(20))


Aperçu des commandes enrichies :
   id_article categorie_produit ss_categorie_produit  revenu
0      PIZ007            Pizzas            Spéciales    12.9
1      PIZ061            Pizzas            Spéciales    13.9
2      PIZ077            Pizzas           Classiques    12.5
3      PIZ011            Pizzas              Sucrées     8.5
4      PIZ106            Pizzas            Spéciales    16.5
5      PIZ064            Pizzas           Gourmandes    15.9
6      SUP038               NaN                  NaN     0.5
7      PIZ044            Pizzas              Sucrées     9.9
8      PIZ083            Pizzas            Spéciales     0.0
9      SUP022               NaN                  NaN     0.5
10     PIZ073            Pizzas            Spéciales    13.9
11     PIZ086            Pizzas           Gourmandes    15.9
12     SUP035               NaN                  NaN     0.5
13     PIZ105            Pizzas            Spéciales    14.5
14     PIZ080            Pizzas            Spéciale

In [25]:
colonnes_sscat = ['kiosque_id', 'ss_categorie_produit', 'annee']

nb_commandes_par_ss_cat_annee = commandes_enrichies.groupby(colonnes_sscat).agg(
    nb_commandes_uniques=('num_commande', 'nunique')
)

# Afficher le résultat
print("--- Nombre de Commandes Uniques par Sous-Catégorie et Année ---")
print(nb_commandes_par_ss_cat_annee)

--- Nombre de Commandes Uniques par Sous-Catégorie et Année ---
                                       nb_commandes_uniques
kiosque_id ss_categorie_produit annee                      
k01        Basiques             2023                   1501
                                2024                   1512
                                2025                   1339
           Classiques           2023                   3486
                                2024                   3826
                                2025                   3197
           Gourmandes           2023                   5038
                                2024                   5437
                                2025                   4636
           Kiosquitos           2023                   1094
                                2024                   1275
                                2025                   1065
           Spéciales            2023                   5912
                                2024

In [26]:
df_sscat = pd.DataFrame(nb_commandes_par_ss_cat_annee)
df_sscat.to_excel("export_sscat.xlsx", index=False)

In [27]:
colonnes_cat = ['categorie_produit', 'annee']

nb_commandes_cat_annee = commandes_enrichies.groupby(colonnes_cat).agg(
    nb_commandes_uniques=('num_commande', 'nunique')
)

# Afficher le résultat
print("--- Nombre de Commandes Uniques par Catégorie et Année ---")
print(nb_commandes_cat_annee)

--- Nombre de Commandes Uniques par Catégorie et Année ---
                         nb_commandes_uniques
categorie_produit annee                      
Boissons          2023                   1215
                  2024                   1684
                  2025                   1565
Desserts          2023                    756
                  2024                    980
                  2025                    862
Kiosquitos        2023                   3222
                  2024                   3839
                  2025                   3538
Pizzas            2023                  30562
                  2024                  34032
                  2025                  30326


### Conclusions :
- la baisse du panier moyen provient de la baisse plus importante des pizzas les plus chères (spéciales et gourmandes)
- particulièrement dans le k04

# Analyse du CA des clients identifiables vs non identifiables

In [28]:

#Création nouvelle colonne pour tager les clients identifiables
commandes_enrichies['is_identifiable'] = np.where(
    (commandes_enrichies['client_id'].notna()) & (commandes_enrichies['client_id'] != 0), 
    True, 
    False
)

In [29]:
ventilation_globale = commandes_enrichies.groupby('is_identifiable').agg(
    ca_total=('revenu', 'sum'),
    nb_commandes=('num_commande', 'nunique')
)

print(ventilation_globale)

                  ca_total  nb_commandes
is_identifiable                         
False             552915.6         23210
True             1751648.6         78825


In [30]:
analyse_panier_moyen = commandes_enrichies.groupby(['is_identifiable', 'annee']).agg(
    ca_total=('revenu', 'sum'),
    nb_commandes=('num_commande', 'nunique')
)
print(analyse_panier_moyen)

                        ca_total  nb_commandes
is_identifiable annee                         
False           2023   190299.65          7571
                2024   198225.95          8497
                2025   164390.00          7142
True            2023   576757.30         24759
                2024   617689.95         28291
                2025   557201.35         25775


### Conclusions :
- Entre 2023 et 2025 :
    - baisse des commandes de clients non identifiés avec baisse du panier moyen
    - hausse des commandes de clients identifiables mais baisse du CA car baisse du panier moyen
- le panier moyen des clients non identifiables est plus élevé que celui des client identifiés

# Analyse des Top clients

In [31]:
# 1. Filtrer uniquement les clients identifiables
df_identifiables = commandes_enrichies[commandes_enrichies['is_identifiable'] == True]

# 2. Calculer le CA total par client
ca_par_client = df_identifiables.groupby('client_id')['revenu'].sum().sort_values(ascending=False)

# 3. Définir le seuil des 20% (ou le nombre de clients à conserver)
# Top 20% des identifiables
seuil_top_20 = int(len(ca_par_client) * 0.20)
top_clients_ids = ca_par_client.head(seuil_top_20).index.tolist()

# 4. Créer un tag 'is_top_client' dans le DataFrame original
commandes_enrichies['is_top_client'] = commandes_enrichies['client_id'].isin(top_clients_ids)

print(f"Les Top 20% des clients représentent {len(top_clients_ids)} IDs.")

Les Top 20% des clients représentent 226 IDs.


In [32]:
# Filtrer sur les top clients identifiés
df_top_clients = commandes_enrichies[commandes_enrichies['is_top_client'] == True]

evolution_top_clients = df_top_clients.groupby('annee').agg(
    nb_commandes_top=('num_commande', 'nunique'),
    ca_total_top=('revenu', 'sum')
)

evolution_top_clients['panier_moyen_top'] = evolution_top_clients['ca_total_top'] / evolution_top_clients['nb_commandes_top']

print("\n--- Évolution Annuelle du Segment TOP 20% ---")
print(evolution_top_clients)


--- Évolution Annuelle du Segment TOP 20% ---
       nb_commandes_top  ca_total_top  panier_moyen_top
annee                                                  
2023              21431     500567.25         23.357158
2024              24470     533997.60         21.822542
2025              22331     482348.10         21.599933


### Conclusions :
- 226 clients constituent environ 80% du CA
- entre 2023 et 2025, le nombre de commandes augmente mais le CA diminue en raison d'une baisse du panier moyen

# ANALYSE DES CLIENTS FIDELES
- Problématique : le gérant constate une baisse de la fréquentation des clients

In [33]:
# 1. Définir les dates d'observation (du 01/04/2023 au 01/12/2025)
# 'MS' signifie 'Month Start' (le 1er du mois)
dates_observation = pd.date_range(start='2023-04-01', end='2025-12-01', freq='MS')

liste_resultats = []

print("Début du traitement par Kiosque...")

for date_obs in dates_observation:
    # 1. Fenêtre glissante [J-90, Date_Obs[
    date_debut = date_obs - pd.Timedelta(days=90)
    
    # 2. Filtrer les commandes de la période
    mask = (commandes['date_commande'] >= date_debut) & (commandes['date_commande'] < date_obs)
    df_window = commandes.loc[mask].copy()
    
    # 3. Calculer le nombre de commandes par Client
    fidelite = df_window.groupby('client_id').size().reset_index(name='nb_commandes_90j')
    
    # Filtrer uniquement les fidèles (>= 3 commandes)
    clients_fideles_ids = fidelite[fidelite['nb_commandes_90j'] >= 3]['client_id']
    
    # Si on a des clients fidèles ce mois-ci...
    if not clients_fideles_ids.empty:
        # On ne garde que les commandes de ces clients fidèles pour déterminer leur kiosque
        commandes_fideles = df_window[df_window['client_id'].isin(clients_fideles_ids)]
        
        # --- 4. DÉTERMINATION DU KIOSQUE DOMINANT ---
        # On trie par date (descendant) pour privilégier le plus récent en cas d'égalité de volume
        commandes_fideles = commandes_fideles.sort_values('date_commande', ascending=False)
        
        # On compte les commandes par Client ET par Kiosque
        kiosque_stats = commandes_fideles.groupby(['client_id', 'kiosque_id']).size().reset_index(name='count')
        
        # On trie pour avoir le kiosque le plus fréquent en premier pour chaque client
        kiosque_stats = kiosque_stats.sort_values(['client_id', 'count'], ascending=[True, False])
        
        # On dédoublonne pour ne garder que la 1ère ligne (le kiosque majoritaire) par client
        kiosque_dominant = kiosque_stats.drop_duplicates(subset=['client_id'], keep='first')[['client_id', 'kiosque_id']]
        
        # --- 5. Finalisation ---
        # On ajoute la date d'observation
        kiosque_dominant['date_observation'] = date_obs
        
        liste_resultats.append(kiosque_dominant)

# --- Consolidation ---
if liste_resultats:
    df_fidelite_kiosque = pd.concat(liste_resultats, ignore_index=True)
    
    print("\n--- Aperçu (Client affecté à son kiosque du moment) ---")
    print(df_fidelite_kiosque.head())


Début du traitement par Kiosque...

--- Aperçu (Client affecté à son kiosque du moment) ---
  client_id kiosque_id date_observation
0    c00016        k04       2023-04-01
1    c00052        k04       2023-04-01
2    c00063        k02       2023-04-01
3    c00068        k04       2023-04-01
4    c00074        k02       2023-04-01


In [34]:
   # --- Analyse Croisée : Évolution par Kiosque ---
evolution_par_kiosque = df_fidelite_kiosque.groupby(['date_observation', 'kiosque_id']).size().unstack(fill_value=0)
    
print("\n--- Évolution des Clients Fidèles par Kiosque ---")
print(evolution_par_kiosque)

evolution_par_kiosque.to_excel("export_clients_fideles.xlsx")


--- Évolution des Clients Fidèles par Kiosque ---
kiosque_id        k01  k02  k03  k04
date_observation                    
2023-04-01        152  183    0  114
2023-05-01        172  162    0  129
2023-06-01        196  159    0  137
2023-07-01        207  145    0  141
2023-08-01        208  144    0  147
2023-09-01        217  137    0  141
2023-10-01        216  140    0  130
2023-11-01        214  145    0  126
2023-12-01        205  160    0  121
2024-01-01        197  167    0  131
2024-02-01        172  174    0  128
2024-03-01        166  188    0  121
2024-04-01        190  186    0  118
2024-05-01        209  169    0  123
2024-06-01        211  165    0  122
2024-07-01        187  182    0  129
2024-08-01        189  190    0  128
2024-09-01        202  180    0  141
2024-10-01        223  164    0  128
2024-11-01        213  179    0  118
2024-12-01        218  162    0  118
2025-01-01        196  157    0  141
2025-02-01        198  148    0  137
2025-03-01        190  1

In [35]:
#Script pour calcul du taux d'attrition

# --- Hypothèses ---
# 1. 'df_fidelite_kiosque' est le résultat du script précédent (colonnes : client_id, kiosque_id, date_observation)
# 2. 'commandes' est votre base de commandes (avec date_commande, client_id)

liste_attrition = []

# On récupère toutes les dates d'observation (les 1er du mois) présentes dans votre étude
dates_uniques = sorted(df_fidelite_kiosque['date_observation'].unique())

print("Calcul de l'attrition en cours...")

for date_debut_mois in dates_uniques:
    # 1. Définir la période de "surveillance" (le mois courant complet)
    # Ex: Si date_debut_mois = 01/04/2023, on regarde jusqu'au 01/05/2023 exclus
    date_fin_mois = date_debut_mois + pd.DateOffset(months=1)
    
    # 2. Isoler les clients fidèles identifiés à ce début de mois
    # On fait une copie pour éviter les avertissements (SettingWithCopyWarning)
    fideles_du_mois = df_fidelite_kiosque[
        df_fidelite_kiosque['date_observation'] == date_debut_mois
    ].copy()
    
    if fideles_du_mois.empty:
        continue

    # 3. Identifier QUI a commandé durant ce mois parmi TOUS les clients
    commandes_du_mois = commandes[
        (commandes['date_commande'] >= date_debut_mois) & 
        (commandes['date_commande'] < date_fin_mois)
    ]
    clients_actifs_ids = set(commandes_du_mois['client_id'].unique())
    
    # 4. Taguer chaque client fidèle : a-t-il commandé ou non ?
    # La colonne 'is_perdu' est Vraie si le client N'EST PAS dans les actifs
    fideles_du_mois['is_perdu'] = ~fideles_du_mois['client_id'].isin(clients_actifs_ids)
    
    # 5. Agréger les résultats par Kiosque
    synthese_mois = fideles_du_mois.groupby(['date_observation', 'kiosque_id']).agg(
        nb_fideles_debut=('client_id', 'count'),
        nb_perdus_fin_mois=('is_perdu', 'sum') # La somme des True donne le nombre de perdus
    ).reset_index()
    
    # Calcul du taux
    synthese_mois['taux_attrition'] = (synthese_mois['nb_perdus_fin_mois'] / synthese_mois['nb_fideles_debut']) * 100
    
    liste_attrition.append(synthese_mois)

# --- Consolidation ---
df_attrition = pd.concat(liste_attrition, ignore_index=True)

# Mise en forme pour l'affichage
pd.options.display.float_format = '{:.2f}%'.format # Format pourcentage
print("\n--- Taux d'Attrition Mensuel des Clients Fidèles par Kiosque ---")
print(df_attrition.head(10))

# --- Optionnel : Vue globale (tous kiosques confondus) ---
df_attrition_global = df_attrition.groupby('date_observation').agg({
    'nb_fideles_debut': 'sum',
    'nb_perdus_fin_mois': 'sum'
}).reset_index()
df_attrition_global['taux_attrition_global'] = (df_attrition_global['nb_perdus_fin_mois'] / df_attrition_global['nb_fideles_debut']) * 100

print("\n--- Évolution Globale du Taux d'Attrition ---")
print(df_attrition_global.head())

df_attrition.to_excel("export_attrition.xlsx")

Calcul de l'attrition en cours...

--- Taux d'Attrition Mensuel des Clients Fidèles par Kiosque ---
  date_observation kiosque_id  nb_fideles_debut  nb_perdus_fin_mois  \
0       2023-04-01        k01               152                  34   
1       2023-04-01        k02               183                  38   
2       2023-04-01        k04               114                  33   
3       2023-05-01        k01               172                  51   
4       2023-05-01        k02               162                  41   
5       2023-05-01        k04               129                  41   
6       2023-06-01        k01               196                  54   
7       2023-06-01        k02               159                  52   
8       2023-06-01        k04               137                  50   
9       2023-07-01        k01               207                  57   

   taux_attrition  
0          22.37%  
1          20.77%  
2          28.95%  
3          29.65%  
4          25.31% 